File: Breadth-First Search/String/LET_no773_Sliding Puzzle.cpp
class Solution {
public:
    int slidingPuzzle(vector<vector<int>>& board) {
        int steps = 0;
        string dest = "123450";
        vector<vector<int>> next_move = {{1,3}, {0,2,4}, {1,5}, {0,4}, {1,3,5}, {2,4}};
        /*
        0 | 1 | 2
        ---------
        3 | 4 | 5
        */

        string start = "";
        for (auto& row : board) {
            for (int num : row)
                start += to_string(num);
        }

        unordered_set<string> visited({start});
        queue<string> q;
        q.push(start);

        while (!q.empty()) {
            int n = q.size();
            while (n-- > 0) {
                string cur = q.front();
                q.pop();

                if (cur == dest)
                    return steps;

                int idx = cur.find("0");
                for (int idx2 : next_move[idx]) {
                    swap(cur[idx], cur[idx2]);
                    if (!visited.count(cur)) {
                        visited.insert(cur);
                        q.push(cur);
                    }
                    swap(cur[idx], cur[idx2]);
                }
            }
            
            steps++;
        }

        return -1;
    }
};

File: Breadth-First Search/String/LET_no773_Sliding Puzzle.py
class Solution:
    def slidingPuzzle(self, board: List[List[int]]) -> int:
        dest = str([[1,2,3],[4,5,0]])
        queue = collections.deque([])
        m, n = 2, 3
        for i in range(m):
            for j in range(n):
                if board[i][j] == 0:
                    queue.append((copy.deepcopy(board), i, j))
                    break
                    
        steps = 0
        visited = {str(board)}
        while queue:
            for _ in range(len(queue)):
                cur, i, j = queue.popleft()
                if str(cur) == dest:
                    return steps
                
                for di, dj in [(0, 1), (1, 0), (-1, 0), (0, -1)]:
                    i2, j2 = i + di, j + dj
                    if 0 <= i2 < m and 0 <= j2 < n:
                        cur[i][j], cur[i2][j2] = cur[i2][j2], cur[i][j]
                        key = str(cur)
                        if key not in visited:
                            visited.add(key)
                            queue.append((copy.deepcopy(cur), i2, j2))
                        cur[i][j], cur[i2][j2] = cur[i2][j2], cur[i][j]
            steps += 1
        
        return -1